src/com/xilinx/rapidwright/design/DesignTools.java

```diff
@@ -3862,18 +3883,68 @@ public static void prohibitPartialHalfSlices(Design design) {
                 }
             }
 
+            // LUTs appear before their corresponding FFs in the iteration sequence so we
+            // can depend on any vacant LUT BELs being populated by the time we need to
+            // evaluate action for the FF BELs
             for (BEL bel : si.getSite().getBELs()) {
                 if (bel.getBELClass() == BELClass.BEL && si.getCell(bel) == null) {
                     Boolean isTop = isUltraScaleSliceTop(bel);
                     if (isTop != null) {
                         if ((isTop && topUsed) || (!isTop && bottomUsed)) {
-                            bels.add(si.getSiteName() + "/" + bel.getName());
+                            if (useDummyCells) {
+                                if (bel.isLUT()) {
+                                    // Being used as a physical static source
+                                    Net lutOutputNet = si
+                                            .getNetFromSiteWire(LUTTools.getLUTOutputPin(bel).getSiteWireName());
+                                    if (lutOutputNet != null) {
+                                        continue;
+                                    }
+
+                                    if (bel.getName().charAt(1) == '5') {
+                                        Net net = si.getNetFromSiteWire(bel.getName().charAt(0) + "6");
+                                        if (net != null) {
+                                            continue;
+                                        }
+                                    }
+
+                                    String name = "dummy_lut_" + dummyCount++;
+                                    Cell c = design.createAndPlaceCell(top, name, Unisim.LUT1,
+                                            si.getSiteName() + "/" + bel.getName());
+                                    c.getEDIFCellInst().addProperty("INIT", "1");
+                                    c.getEDIFCellInst().addProperty(EDIFTools.DONT_TOUCH, value);
+                                    c.removePinMapping("A6");
+                                    c.removePinMapping("A5");
+                                    c.setBELFixed(true);
+                                    c.setSiteFixed(true);
+                                    boolean connected = false;
+                                    for (int i = 1; i < 5; i++) {
```

Comment:
Am I right in understanding that this block of code looks at `A1` to `A4` and attempts finds a pin with a static source, and connects that to `I0` of the newly created `LUT1`.
If no static source is found (i.e. when neither the `?6LUT` nor `?5LUT` BEL is occupied and thus no LUT inputs are used?) then this cell is removed again (and this BEL is not prohibited anymore?)

Reply:
```suggestion
                                    for (int i = 1; i <= 5; i++) {
```
Yes (although, I think it should be from `A1` to `A5`).  We only want to add a `LUT1` to block the cell if there is an unused pin.  If all the pins are already used, then there is no need to add a `LUT1` as any cell that could be placed there would have to have a strict subset of connections that are already being made and thus wouldn't add any changes to the routing burden of the design (at least on the input cones).

Reply:
Okay, I think I understand now. Find the first unused pin or the first pin with a static source, and map it to `I0`. If there are no such pins, this means that all pins are used and the placer would not put something there anyway.